miprometheus/problems/question_context_to_class/babiqa_dataset_single_question.py

```diff
@@ -0,0 +1,514 @@
+
+        k = 0
+        for i, word in enumerate(current_question[0].split(' ')):
+            if word == '_':
+                mask[i] = 1
+                target[i] = answer[k]
+                k=k+1
+
+        #make a dictionnary with all the outputs
+        data_dict = self.create_data_dict()
+        data_dict['sequences'] = story
+        data_dict['targets'] = target
+        data_dict['current_question'] = current_question
+        data_dict['masks'] = mask
+
+        #return the final DataDict
+        return data_dict
+
+    def collate_babi(self, batch):
+
+        """            
+               Collate method that create batch from samples.
+               :param batch.
+               :return: return {'sequence': sequence, 'targets': targets, 'current_question': current_question, "mask": mask} 
+               """
+        # get sizes
+        context_length = max(d["sequences"].shape[0] for d in batch)
+        answer_length = max(d["targets"].shape[0] for d in batch)
+        batch_size = len(batch)
+        word_size = batch[0]["sequences"].shape[-1]
+
+        # create placeholders
+        sequences = torch.zeros((batch_size, context_length, word_size)).type(AppState().dtype)
+
+        targets = torch.zeros((batch_size, answer_length)).type(AppState().LongTensor)
+        mask = torch.zeros((batch_size, answer_length)).type(AppState().ByteTensor)
+
+        # padded data
+        current_question = []
+        for i, d in enumerate(batch):
+            c_shape = d["sequences"].shape
+            a_shape = d["targets"].shape
+            sequences[i, :c_shape[0], :c_shape[1]] = d["sequences"]
+            targets[i, :a_shape[0]] = d["targets"]
+            mask[i, :a_shape[0]] = d["masks"]
+            current_question.append(d["current_question"])
+
+            # make a dictionnary with all the outputs
+            data_dict = self.create_data_dict()
+            data_dict['sequences'] = sequences
+            data_dict['targets'] = targets
+            data_dict['current_question'] = current_question
+            data_dict['masks'] = mask
+
+        # return the fina DataDict
+        return data_dict
+
+    def evaluate_loss(self, data_dict, logits):
+        """ Calculates accuracy equal to mean number of correct predictions in a given batch.
+        WARNING: Applies mask to both logits and targets!
+
+        :param data_dict: DataDict({'sequences', 'sequences_length', 'targets', 'mask'}).
+
+        :param logits: Predictions being output of the model.
+
+        """
+        # Check if mask should be is used - if so, use the correct loss
+        # function.
+        if self.use_mask:
+            loss = self.loss_function(
+                logits, data_dict['targets'], data_dict['masks'])
+        else:
+            pred = logits.transpose(1, 2)
+
+            loss = self.loss_function(pred, data_dict['targets'], data_dict['masks'])
+
+        return loss
+
+
+    def build_dictionaries_one_hot(self):
+
```

Comment:
Description is inconsistent with the name of the function. Is it really creating word embeddings OR is it building dictionaries from all words present in that stories/questions??

Why do we need two methods for it (with one hot and without)?

Reply:
We could keep only one. There is a boolean variable self.embedding_type that decides if you want One hot or embedding or not. I though this is a nice feature so I kept it .